src/main/scala/is/hail/linalg/GridPartitioner.scala

```diff
@@ -39,4 +40,81 @@ case class GridPartitioner(blockSize: Int, nRows: Long, nCols: Long) extends Par
   }
   
   def transpose: GridPartitioner = GridPartitioner(this.blockSize, this.nCols, this.nRows)
+  
+  // returns all blocks intersecting the diagonal band consisting of all entries
+  //   with -lowerBandwidth <= (colIndex - rowIndex) <= upperBandwidth
+  def bandedBlocks(lowerBandwidth: Long, upperBandwidth: Long): Array[Int] = {
+    require(lowerBandwidth >= 0 && upperBandwidth >= 0)
+    
+    val lowerBlockBandwidth = blockIndex(lowerBandwidth + blockSize - 1)
+    val upperBlockBandwidth = blockIndex(upperBandwidth + blockSize - 1)
+
+    val blocks = new ArrayBuilder[Int]
+    
+    var j = 0
+    while (j < nBlockCols) {
+      val offset = j * nBlockRows
+      var i = (j - upperBlockBandwidth) max 0
+      while (i <= ((j + lowerBlockBandwidth) min (nBlockRows - 1))) {
+        blocks += offset + i
+        i += 1
+      }
+      j += 1
+    }
+    
+    blocks.result()
+  }
+  
+  // returns all blocks intersecting the rectangle [firstRow, lastRow] x [firstCol, lastCol]
+  def rectangularBlocks(firstRow: Long, lastRow: Long, firstCol: Long, lastCol: Long): Array[Int] = {
+    require(firstRow >= 0 && firstRow <= lastRow && lastRow <= nRows)
+    require(firstCol >= 0 && firstCol <= lastCol && lastCol <= nCols)
+    
+    val firstBlockRow = blockIndex(firstRow)
+    val lastBlockRow = blockIndex(lastRow)
+    val firstBlockCol = blockIndex(firstCol)
+    val lastBlockCol = blockIndex(lastCol)
+
+    val blocks = new Array[Int]((lastBlockRow - firstBlockRow + 1) * (lastBlockCol - firstBlockCol + 1))
+    
+    var k = 0
+    var j = firstBlockCol
+    while (j <= lastBlockCol) {
+      val offset = j * nBlockRows
+      var i = firstBlockRow
+      while (i <= lastBlockRow) {
+        blocks(k) = offset + i
+        k += 1
+        i += 1
+      }
+      j += 1
+    }
```

Comment:
Likewise, this could all be condensed to
```scala
(for { j <- firstBlockCol until lastBlockCol
       i <- firstBlockRow until lastBlockRow
} yield (j * nBlockRows) + i ).toArray
```